File: android/app/src/main/kotlin/com/essama/dresscode/ui/ecrans/Commandes.kt
package com.essama.dresscode.ui.ecrans

import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.FilterChip
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.ui.graphics.Color
import androidx.navigation.NavHostController
import com.essama.dresscode.charte.Espace
import com.essama.dresscode.charte.IconeSymbole
import com.essama.dresscode.charte.Icones
import com.essama.dresscode.charte.Taille
import com.essama.dresscode.metier.Commande
import com.essama.dresscode.metier.Ligne
import com.essama.dresscode.metier.Statut
import com.essama.dresscode.metier.delai
import com.essama.dresscode.metier.etat
import com.essama.dresscode.metier.montant
import com.essama.dresscode.ui.CarteLien
import com.essama.dresscode.ui.EtatVide
import com.essama.dresscode.ui.ModeleVue
import com.essama.dresscode.ui.Route
import com.essama.dresscode.ui.Vignette

/*
 * Les filtres reprennent exactement les blocs d'« Aujourd'hui » :
 * appuyer sur « 2 commandes en retard » doit amener sur ces deux
 * commandes-la, pas sur une liste ou il faut les rechercher.
 */

private data class Filtre(
    val cle: String,
    val libelle: String,
    val garde: (Ligne) -> Boolean,
)

private val filtres = listOf(
    Filtre("en_cours", "En cours") { it.commande.statut != Statut.LIVREE },
    Filtre("retard", "En retard") { it.etat.enRetard },
    Filtre("aujourdhui", "Aujourd’hui") { it.etat.livraisonAujourdhui },
    Filtre("a_commencer", Statut.A_COMMENCER.libelle) { it.commande.statut == Statut.A_COMMENCER },
    Filtre("en_confection", Statut.EN_CONFECTION.libelle) { it.commande.statut == Statut.EN_CONFECTION },
    Filtre("prete", Statut.PRETE.libelle) { it.commande.statut == Statut.PRETE },
    Filtre("impayees", "Impayées") { it.etat.reste > 0 },
    Filtre("historique", "Historique") { it.commande.statut == Statut.LIVREE },
    Filtre("toutes", "Toutes") { true },
)

@Composable
fun EcranCommandes(
    modeleVue: ModeleVue,
    navigation: NavHostController,
    filtreInitial: String,
) {
    val commandes by modeleVue.commandes.collectAsState()
    val clients by modeleVue.clients.collectAsState()
    var choisi by remember { mutableStateOf(filtreInitial) }

    val lignes = commandes.map { Ligne(it, it.etat()) }
    val filtre = filtres.firstOrNull { it.cle == choisi } ?: filtres.first()

    /* Le plus urgent en haut, toujours : une liste triee par date de
       creation obligerait a la parcourir en entier pour trouver ce
       qui brule. */
    val visibles = if (filtre.cle == "historique") {
        /* L'historique se lit a l'envers du reste : ce qu'on cherche,
           c'est la derniere robe faite pour cette cliente, pas la
           premiere de l'annee. */
        lignes.filter(filtre.garde)
            .sortedWith(
                compareByDescending<Ligne> { it.commande.dateLivraison }
                    .thenByDescending { it.commande.livreeLe ?: 0L },
            )
    } else {
        lignes.filter(filtre.garde)
            .sortedWith(
                compareBy<Ligne> { if (it.commande.statut == Statut.LIVREE) 1 else 0 }
                    .thenBy { it.etat.joursRestants },
            )
    }

    LazyColumn(
        contentPadding = PaddingValues(
            start = Espace.quatre, end = Espace.quatre,
            top = Espace.quatre, bottom = Espace.seize,
        ),
        verticalArrangement = Arrangement.spacedBy(Espace.trois),
    ) {
        item {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .horizontalScroll(rememberScrollState())
                    .padding(bottom = Espace.deux),
                horizontalArrangement = Arrangement.spacedBy(Espace.deux),
            ) {
                filtres.forEach { candidat ->
                    val nombre = lignes.count(candidat.garde)
                    FilterChip(
                        selected = candidat.cle == choisi,
                        onClick = { choisi = candidat.cle },
                        label = {
                            Text(
                                if (nombre > 0) "${candidat.libelle}  $nombre" else candidat.libelle,
                            )
                        },
                    )
                }
            }
        }

        if (visibles.isEmpty()) {
            item {
                EtatVide(icone = Icones.Checkroom, titre = messageVide(filtre.cle))
            }
        }

        items(visibles, key = { it.commande.id }) { ligne ->
            val client = clients.firstOrNull { it.id == ligne.commande.clientId }
            val photo = ligne.commande.photo?.let { modeleVue.depot.photos.fichier(it) }
            CarteLien(
                titre = ligne.commande.modeleNom,
                detail = listOfNotNull(
                    client?.nom ?: "Cliente supprimée",
                    /* La photo remplace la pastille : le statut, qu'elle
                       portait, revient ici en toutes lettres. */
                    if (photo != null) ligne.commande.statut.libelle else null,
                    if (ligne.etat.reste > 0) "reste ${montant(ligne.etat.reste)}" else null,
                ).joinToString(" · "),
                debut = {
                    Vignette(
                        fichier = photo,
                        description = ligne.commande.modeleNom,
                    ) { JetonStatut(ligne.commande.statut) }
                },
                fin = { Echeance(ligne) },
                surClic = { navigation.navigate(Route.commande(ligne.commande.id)) },
            )
        }
    }
}

/* La couleur seule ne dit rien : l'icone porte le sens, et le
   libelle complet est sur la fiche. */
@Composable
private fun JetonStatut(statut: Statut) {
    val fond = when (statut) {
        Statut.A_COMMENCER -> MaterialTheme.colorScheme.surfaceVariant
        Statut.EN_CONFECTION -> MaterialTheme.colorScheme.secondaryContainer
        Statut.PRETE -> MaterialTheme.colorScheme.primaryContainer
        Statut.LIVREE -> MaterialTheme.colorScheme.tertiaryContainer
    }
    val encre = when (statut) {
        Statut.A_COMMENCER -> MaterialTheme.colorScheme.onSurfaceVariant
        Statut.EN_CONFECTION -> MaterialTheme.colorScheme.onSecondaryContainer
        Statut.PRETE -> MaterialTheme.colorScheme.onPrimaryContainer
        Statut.LIVREE -> MaterialTheme.colorScheme.onTertiaryContainer
    }
    Box(
        modifier = Modifier.size(40.dp).background(fond, CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        IconeSymbole(icone = iconeDe(statut), taille = Taille.petite, couleur = encre)
    }
}

@Composable
private fun Echeance(ligne: Ligne) {
    val couleur: Color = when {
        ligne.commande.statut == Statut.LIVREE -> MaterialTheme.colorScheme.onSurfaceVariant
        ligne.etat.enRetard -> MaterialTheme.colorScheme.error
        ligne.etat.livraisonAujourdhui || ligne.etat.aCommencer -> MaterialTheme.colorScheme.primary
        else -> MaterialTheme.colorScheme.onSurfaceVariant
    }
    Text(
        text = if (ligne.commande.statut == Statut.LIVREE) {
            "Livrée"
        } else {
            delai(ligne.commande.dateLivraison)
        },
        style = MaterialTheme.typography.labelLarge,
        color = couleur,
    )
}

internal fun iconeDe(statut: Statut) = when (statut) {
    Statut.A_COMMENCER -> Icones.ContentCut
    Statut.EN_CONFECTION -> Icones.Iron
    Statut.PRETE -> Icones.CheckCircle
    Statut.LIVREE -> Icones.Inventory2
}

private fun messageVide(filtre: String) = when (filtre) {
    "en_cours" -> "Aucune commande en cours."
    "retard" -> "Aucun retard."
    "aujourdhui" -> "Aucune livraison aujourd’hui."
    "impayees" -> "Tout est encaissé."
    "historique" -> "Aucune commande livrée pour le moment."
    "toutes" -> "Aucune commande enregistrée."
    else -> "Aucune commande à ce stade."
}

File: app/js/vues/commandes.js
/* ============================================================
   LISTE DES COMMANDES
   ------------------------------------------------------------
   Les filtres reprennent exactement les blocs d'« Aujourd'hui » :
   cliquer sur « 2 commandes en retard » doit amener sur ces deux
   commandes-la, pas sur une liste ou il faut les rechercher.
   ============================================================ */

import { lireTout } from '../donnees.js';
import { STATUTS, delai, etat, montant, reste } from '../metier.js';
import { etatVide, icone, txt } from '../interface.js';
import { aller } from '../routeur.js';

const FILTRES = [
  { cle: 'en_cours', libelle: 'En cours' },
  { cle: 'retard', libelle: 'En retard' },
  { cle: 'aujourdhui', libelle: 'Aujourd’hui' },
  { cle: 'a_commencer', libelle: STATUTS.a_commencer.libelle },
  { cle: 'en_confection', libelle: STATUTS.en_confection.libelle },
  { cle: 'prete', libelle: STATUTS.prete.libelle },
  { cle: 'impayees', libelle: 'Impayées' },
  { cle: 'toutes', libelle: 'Toutes' },
];

const GARDE = {
  en_cours: (c) => c.commande.statut !== 'livree',
  retard: (c) => c.etat.enRetard,
  aujourdhui: (c) => c.etat.livraisonAujourdhui,
  a_commencer: (c) => c.commande.statut === 'a_commencer',
  en_confection: (c) => c.commande.statut === 'en_confection',
  prete: (c) => c.commande.statut === 'prete',
  impayees: (c) => c.etat.reste > 0,
  toutes: () => true,
};

export async function vueCommandes(_, recherche) {
  const filtre = FILTRES.some((f) => f.cle === recherche?.get('filtre'))
    ? recherche.get('filtre')
    : 'en_cours';

  const [commandes, clients] = await Promise.all([lireTout('commandes'), lireTout('clients')]);
  const parId = new Map(clients.map((c) => [c.id, c]));

  const enrichies = commandes.map((commande) => ({ commande, etat: etat(commande) }));
  const visibles = enrichies.filter(GARDE[filtre]).sort(trier);

  const section = document.createElement('div');
  section.innerHTML = `
    <div class="filtres" role="tablist">
      ${FILTRES.map((f) => {
        const nombre = enrichies.filter(GARDE[f.cle]).length;
        return `
          <button class="md-chip md-chip-filter ${f.cle === filtre ? 'md-chip-selected' : ''}"
                  role="tab" aria-selected="${f.cle === filtre}" data-filtre="${f.cle}">
            ${txt(f.libelle)}${nombre ? ` <span class="md-badge md-badge-neutral">${nombre}</span>` : ''}
          </button>`;
      }).join('')}
    </div>
    ${visibles.length
      ? visibles.map((x) => carte(x, parId.get(x.commande.clientId))).join('')
      : etatVide({ icone: 'checkroom', titre: messageVide(filtre) })}`;

  section.addEventListener('click', (evenement) => {
    const chip = evenement.target.closest('[data-filtre]');
    if (chip) {
      aller(`/commandes?filtre=${chip.dataset.filtre}`);
      return;
    }
    const carte = evenement.target.closest('[data-commande]');
    if (carte) aller(`/commande/${carte.dataset.commande}`);
  });

  return section;
}

/* Le plus urgent en haut, toujours : une liste triee par date de
   creation obligerait a la parcourir en entier pour trouver ce
   qui brule. */
function trier(a, b) {
  const rang = (x) => (x.commande.statut === 'livree' ? 1 : 0);
  if (rang(a) !== rang(b)) return rang(a) - rang(b);
  return (a.etat.restants ?? 9999) - (b.etat.restants ?? 9999);
}

function carte({ commande, etat: situation }, client) {
  const du = reste(commande);
  const statut = STATUTS[commande.statut];
  const classeEcheance = situation.enRetard
    ? 'echeance-retard'
    : situation.livraisonAujourdhui || situation.aCommencer ? 'echeance-proche' : '';

  return `
    <button class="carte-lien" data-commande="${txt(commande.id)}">
      <span class="jeton-statut" data-statut="${txt(commande.statut)}" aria-hidden="true">
        ${icone(statut.icone, { taille: 20 })}
      </span>
      <span class="carte-corps">
        <span class="carte-titre">${txt(commande.modeleNom)}</span>
        <span class="carte-detail">${txt(client?.nom || 'Cliente supprimée')}${
          du > 0 ? ` · reste ${txt(montant(du))}` : ''}</span>
      </span>
      <span class="echeance ${classeEcheance}">
        ${commande.statut === 'livree' ? 'Livrée' : txt(delai(commande.dateLivraison))}
      </span>
    </button>`;
}

const messageVide = (filtre) =>
  ({
    en_cours: 'Aucune commande en cours.',
    retard: 'Aucun retard.',
    aujourdhui: 'Aucune livraison aujourd’hui.',
    impayees: 'Tout est encaissé.',
    toutes: 'Aucune commande enregistrée.',
  })[filtre] || `Aucune commande à ce stade.`;
